src/backend/lib/upgrade-storage.js

```diff
@@ -42,13 +43,44 @@ class Storage {
     log("..done. recreating temp directory..");
     mkdirp.sync(this.tmpdir);
     log("..done");
-
+    this.lock = new RWLock();
+    log("deleting old apks");
+    this.lock.writeLock(release => {
+      function done(err) {
+        if (err) throw new Error(err);
+        log("..done deleting old apks");
+        release();
+      }
+      this.init(done);
+    });
     opts = opts || {};
     this.targetPlatform = opts.platform || "android";
     this.targetArch = opts.arch || "arm64-v8a";
     this.version = opts.version || "0.0.0";
   }
 
+  init(cb) {
+    // TODO: async version of the mkdir/rimraf stuff we already have
+    this.localUpgrades.get((err, options) => {
+      if (err) return cb(err);
+      const toDelete = options.filter(
+        option => !this.isNewerThanCurrentApk(option)
+      );
+      const toKeep = options.filter(option =>
+        this.isNewerThanCurrentApk(option)
+      );
+      foreachAsync(
+        toDelete,
+        (option, cb) => {
+          rimraf(option.filename, cb);
+        },
+        _ => {
```

Comment:
@gmaclennan Thoughts on whether we should error if some of the deletions fail?

I'm thinking it might be safest to add the files that we failed to delete to the `toKeep` list, so at least they are tracked. Otherwise we could have failed-to-delete entries clogging up disk space that the upgrade system doesn't even have a reference to. So, my suggestion is: don't have `foreachAsync` fail on an error, but rather accumulate failures + report them back, and then add those failed deletions back to `toKeep` so they stay tracked.